src/core/execute.c

```diff
@@ -3277,6 +3286,28 @@ static int exec_child(
                         log_unit_warning(unit, "PrivateNetwork=yes is configured, but the kernel does not support network namespaces, ignoring.");
         }
 
+#if HAVE_APPARMOR
+        if (needs_sandboxing) {
+                if (use_apparmor && context->apparmor_profile) {
+                        r = aa_change_onexec(context->apparmor_profile);
+                        if (r < 0 && !context->apparmor_profile_ignore) {
+                                *exit_status = EXIT_APPARMOR_PROFILE;
+                                return log_unit_error_errno(unit, errno, "Failed to prepare AppArmor profile change to %s: %m", context->apparmor_profile);
+                        }
+                }
+        }
+#endif
+
+        if (needs_sandboxing) {
+                if (context->private_users) {
+                        r = setup_private_users(uid, gid);
```

Comment:
this looks wrong. if we are inside a user ns we might not be able to mount everything we might want to mount. hence this really should be placed after the namespace calls further down, or what am i missing?

Reply:
quite frankly, I am not sure it's worth supporting MountAPIVFS=no at all in combination with PrivateUsers= (when we added it I was aleady arguing in favour of not having MountAPIVFS= at all, and just always doing it).

Hence, maybe the way out here is to imply MountAPIVFS=yes behaviur if privateusers= is set.

Reply:
What do you want to mount? It works fine with BindPaths/BindReadOnlyPaths for me. If someone else tests it, I will be happy.

The idea behind SystemCallFilter is to minimize the attack surface of the kernel. I think, the same can be said about the API filesystems. Or think about /proc/self/exe and runc. There are more and more API filesystems in the kernel. Most of them are used to configure the kernel. Some are used to get information from the kernel. Hardly any daemon really needs them. I use RootDirectory as a whitelist of mounts. I know there is InaccessiblePaths but it requires me to over-mount any of those filesystems. It also requires me to know all those filesystems. With RootDirectory and MountAPIVFS=false I can easly get a minimal environment.

Reply:
Well, think is access to /proc and /sys is cooked at plenty places into glibc and many other libs. So even if you are careful to never access procfs and sysfs you implicitly do, because it happens underneath. Hence I am not sure it's worth to bother, because the Linux API *is* /proc and /sys and suppressing that is likely one step too far...

Reply:
Almost all services for network protocols support chroot. All those chroots are setup without /proc or /sys inside and they work fine for 20+ years. I know that this is a bit different, because if the services do the chroot themselves, they have already initialized themselves before the chroot. But still, I run bird and qemu in such a chroot and it works fine. Qemu needs some additional device files in its chroot environment but this can be done easily with BindPath.

You are certainly correct, that it may cause problems and not every service can be run in its own chroot environment without /proc or /sys but there are service that work this way. So why prevent people from doing this on purpose? If it does not work for a service, it is just not supported by systemd. But if it does work, why not let the people use it? There is no harm in it.

```
[Unit]
Description=Virtual Machine %i

Documentation=man:qemu

ConditionFileIsExecutable=/usr/bin/vm
ConditionFileIsExecutable=/usr/bin/qemu-system-x86_64
ConditionCapability=cap_net_admin

After=apparmor-system@%N.service network-online.target
Wants=apparmor-system@%N.service network-online.target

RequiresMountsFor=/var/lib/machines

[Service]
User=%N
Group=%N
SupplementaryGroups=kvm
DynamicUser=true

RuntimeDirectory=%N %N/chroot
RuntimeDirectoryMode=0700

RootDirectoryStartOnly=true
RootDirectory=%t/%N/chroot
MountAPIVFS=false
BindReadOnlyPaths=/etc/ld.so.cache /lib -/lib64 /usr/lib
BindReadOnlyPaths=/usr/bin/qemu-system-x86_64
BindReadOnlyPaths=/dev/random /dev/urandom
BindPaths=/dev/null
BindPaths=/dev/vhost-net /dev/kvm /dev/net/tun
BindPaths=/proc/sys/net/ipv4/conf
BindPaths=-/var/lib/machines/%i.raw
BindReadOnlyPaths=-/usr/share/qemu -/usr/share/seabios -/usr/share/seavgabios -/usr/share/ipxe -/var/pactoo/repo -/var/portage

Environment="INSTANCE=%i"
Environment="KEYBOARD=de"

ExecStartPre=/usr/bin/vm prepare %i
ExecStartPre=/usr/bin/chown root:root -- %t/%N/chroot
ExecStartPre=/usr/bin/chmod 0755 -- %t/%N/chroot
ExecStart=/usr/bin/qemu-system-x86_64 -no-user-config -nodefaults -readconfig %t/%N/config -fw_cfg opt/vm-utils/config,file=%t/%N/fw_cfg -cpu host,+spec-ctrl -boot order=cn,menu=on -watchdog-action reset -k $KEYBOARD $ARGUMENTS
ExecStopPost=/usr/bin/vm cleanup %i

KillMode=process
KillSignal=SIGINT
TimeoutStopSec=60s
Restart=always
RestartSec=5s

PermissionsStartOnly=true
AppArmorProfile=:system:%N
PrivateUsers=true
CapabilityBoundingSet=
NoNewPrivileges=true
LockPersonality=true
SystemCallArchitectures=native
SystemCallFilter=~@chown @clock @cpu-emulation @debug @keyring @module @mount @obsolete @privileged @raw-io @reboot @resources @setuid @swap
SystemCallErrorNumber=eperm
IPAddressDeny=any
IPAddressAllow=127.0.0.1/8 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 100.64.0.0/10
RestrictAddressFamilies=af_unix af_inet af_inet6 af_netlink
RestrictNamespaces=true
RestrictRealtime=true
MemoryDenyWriteExecute=true
DevicePolicy=closed
DeviceAllow=/dev/kvm rw
DeviceAllow=/dev/net/tun rw
DeviceAllow=/dev/vhost-net rw
RemoveIPC=true

SyslogIdentifier=%N

[Install]
WantedBy=machines.target
```